src/VisualStudio/Core/Def/LanguageService/PackageRegistrationTasks.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService;
+
+using WorkTask = Func<IProgress<ServiceProgressData>, PackageRegistrationTasks, CancellationToken, Task>;
+
+/// <summary>
+/// Provides a mechanism for registering work to be done during package initialization. Work is registered
+/// as either main thread or background thread appropriate. This allows processing of these work items
+/// in a batched manner, reducing the number of thread switches required during the performance sensitive
+/// package loading timeframe.
+/// 
+/// Note that currently the processing of these tasks isn't done concurrently. A future optimization may
+/// allow parallel background thread task execution, or even concurrent main and background thread work.
+/// </summary>
+internal sealed class PackageRegistrationTasks(JoinableTaskFactory jtf)
+{
+    private readonly ConcurrentQueue<WorkTask> _backgroundThreadWorkTasks = [];
+    private readonly ConcurrentQueue<WorkTask> _mainThreadWorkTasks = [];
```

Comment:
might want to mention somewhere that there can't be any dependencies (currently) between main and bg thread tasks.